### ISSUE TITLE
VOLUME contains data added after declaration

### ISSUE THREAD

**Issue by twaugh** [openshift/imagebuilder#17]
I understand this is a known issue, that a Dockerfile like the following will end up creating a volume differently when built using imagebuilder:
```
FROM busybox
RUN mkdir /vol
RUN touch /vol/foo
VOLUME /vol
RUN touch /vol/bar
```

It may be that a volume is carefully set up in a Dockerfile, then some other command is run which -- as a side effect -- modifies that volume content. The author may be using the knowledge that the volume declaration 'freezes' the initial volume content in order to know the side-effect-producing command is safe to run.

I'd like to propose a solution: during the implementation of VOLUME, imagebuilder could export the content of the container and capture the content. This is the author's intended initial content for the volume.

After all the Dockerfile instructions are run, a final set of instructions could restore the content of volumes declared in the Dockerfile to the state they were in at declaration. For instance, saving the current user, adding a binary to clean out volume content (and afterwards, itself), becoming root and running it, then restoring the current user.

**Comment by smarterclayton** [comment]
This is a known issue, there's really no way to fix it without executing things in the container, which we can only do if we know that the image has a shell in it.  So we can only fix part of the problem.

**Comment by smarterclayton** [comment]
How critical is this?

**Comment by twaugh** [comment]
Injecting a static `rm` binary was the proposal I had.

Short of that, if imagebuilder could at least warn when the volume content ends up different than at declaration time, it would at least enable any tooling using it to re-build the container image a different way if it wants full compatibility for VOLUME.

**Comment by smarterclayton** [comment]
Mostly fixed, but see #30 for the gap with owner permissions.  I will add a `--strict-volume-ownership` flag to prevent this accidentally being lost.

**Comment by smarterclayton** [comment]
Added a check in #31 which will fail if RUN follows VOLUME.  ADD following VOLUME is successful.
